### ISSUE TITLE
[DataLoader] Implement Map-style functional DataPipe

### ISSUE THREAD

**Issue by ejguan** [pytorch/pytorch#57031]
## 🚀 Feature
Add map-style Datapipe to support new DataLoader.

## Motivation
From the RFC #49440, you can find lots of issues with the previous DataLoader. So, we are working on a new design of DataLoader to support several features such as multithreading, and to prevent OOM. To support the new design of DataLoader, we have proposed the `IterDataPipe` for the iter-style DataPipe. But, in terms backward-compatibility, 
a list of functional `MapDataPipe` is also needed for users. The difference between iter-style and map-style can be found [here
](https://pytorch.org/docs/stable/data.html#dataset-types)

You can take reference from the implementation of [iter-style DataPipe](https://github.com/pytorch/pytorch/blob/6bbd8ba658d573956744dc68fed4614bd0fbd5d1/torch/utils/data/datapipes/iter/callable.py#L29-L82)

## Note
- path: torch/utils/data/datapipes/map/...
- `__len__` is required for map-style
- `__getitem__` is required for map-style
- Make sure DataPipe lazy-init and loading data lazily
- Make sure DataPipe is pickleable

## List
- [x] ConcatMapDataPipe
- [x] ZipMapDataPipe
- [x] BatchMapDataPipe
- [ ] UnbatchMapDataPipe (Likely do not want it)
- [x] MapMapDataPipe
- [ ] UnZipMapDataPipe (Likely do not want it)
- [x] ShuffleMapDataPipe
- [ ] FilterMapDataPipe (Likely do not want it)
- [ ] ForkMapDataPipe
- [ ] DemuxMapDataPipe (Likely do not want it, length not known in advance)
- [ ] MultiplexerMapDataPipe (Likely do not want it since the functionality doesn't make sense in Map-style? Open to suggestion if there is an actual use case)
- [ ] ... 

Here are the [complete list of `IterDataPipe`](https://pytorch.org/data/main/torchdata.datapipes.iter.html), some of which we may want to create Map-style version:

---
To avoid implementing every single feature twice (one in Iter-style, one in Map-style), our current [guidance](https://pytorch.org/data/main/torchdata.datapipes.map.html) is that users should use the built-in `IterDataPipe` for various functionalities, and convert it to MapDataPipe as needed using [`IterToMapConverter`](https://pytorch.org/data/main/generated/torchdata.datapipes.map.IterToMapConverter.html#torchdata.datapipes.map.IterToMapConverter) or .to_map_datapipe(). For that reason, I don't think we will need additional `MapDataPipe` for now.

Feedback and suggestions are welcome here.

cc @SsnL @VitalyFedyunin @ejguan @NivekT

**Comment by zhuzilin** [comment]
Hi! I wonder if I can help with some of the map stype data pipe? Or they all have been assigned to new on-board members?

**Comment by ejguan** [comment]
@zhuzilin 
Thanks for following up. Feel free to open PR to implement Map-style DP.
Do you mind starting from Zip or Concat? We have a new stack of PRs that are going to slightly change the API of some other functional datapipes.

**Comment by zhuzilin** [comment]
@ejguan That's great! I'll try from Concat~

**Comment by KinglittleQ** [comment]
@ejguan Could I implement BatchMapDataPipe for you?

**Comment by ejguan** [comment]
@KinglittleQ Thank you for taking interest. Please feel free to open PR and request review.

**Comment by NivekT** [comment]
As an update to this issue, based on discussion elsewhere, it is not clear that we want to have the functions `unbatch` and `filter`, because those operations cannot be performed **lazily** without strong assumptions about the shape of the inputs.

For instance, `unbatch`, if done lazily, may requires assumption about each batch having the same size. `filter` requires knowing the number of elements that match the requirement prior.

We do not have a `unzip` DataPipe for `IterDataPipe` and that is likely not needed either.

**Comment by josiahls** [comment]
@ejguan I know I already tagged you in that issue however just so others are aware, I'm tackling  DemuxMapDataPipe /
 MultiplexerMapDataPipe in https://github.com/pytorch/data/issues/310

**Comment by ykabusalah** [comment]
Hello, I see that this issue has still not been addressed. Would I be able to tackle this along with the two other issues I asked for? 

Thank you!

**Comment by NivekT** [comment]
> Hello, I see that this issue has still not been addressed. Would I be able to tackle this along with the two other issues I asked for?
> 
> Thank you!

Hi, what are DataPipes are you looking for and what are the two other issues?

We are in the process of providing more guidance to users about `MapDataPipe`. In general, rather than duplicating the same functionalities twice for `Iter` and `Map`, we believe it is better to use `IterDataPipe` and provide a converter to convert `IterDataPipe` to `MapDataPipe` when necessary. We are open to feedback in this regard.

We may add a few `MapDataPipes` where the operations can be lazily done and `__len__` can be known in advance. If you can tell us about your use cases and how `IterDataPipe` isn't sufficient, we will definitely take that into consideration or provide a pattern that may work for you.

**Comment by ykabusalah** [comment]
I am looking at two other good first issues that I have commented on, so please ignore that part. 

And what do you mean by what are DataPipes are you looking for? @NivekT

**Comment by NivekT** [comment]
> I am looking at two other good first issues that I have commented on, so please ignore that part.
> 
> And what do you mean by what are DataPipes are you looking for? @NivekT

I thought you wanted us to add specific `MapDataPipe` due to your use case, but if that is not the case, nevermind.

**Comment by ykabusalah** [comment]
You certainly could I just didn't understand initially. But could I tackle this issue then? @NivekT

**Comment by NivekT** [comment]
Thanks for your interest. There are several open PRs related to this issue so we are good for now.

There are [other good first issues here for TorchData](https://github.com/pytorch/data/contribute), let us know if you are interested in working on any of those that do not have existing open PRs.

**Comment by PhilMacKay** [comment]
Hello! I was reading through the documentation because we are in the process of implementing data pipes, and I came upon the link to this issue thread.

One tiny suggestion I would have to make is this: when searching through the API, it's hard to know that `map.SequenceWrapper` accepts a `collections.abc.Mapping` (ex: `dict`). Its name makes it sound like it only accepts sub-classes of `collections.abc.Sequence` (ex: `list`). The code example makes it clear that it accepts dictionaries, but the name and description do not.

I don't have a single suggestion, because it doesn't seem obvious between:
1. Rename `map.SequenceWrapper` to `map.CollectionWrapper` (does it accept any subclass of `collections.abc.Collection` such as `set`?)
2. Create a `map.MappingWrapper` that is just an alias to `SequenceWrapper`
3. Create a new `map.MappingWrapper` and make it such that `SequenceWrapper` no longer accepts `collections.abc.Mapping`.

**Comment by ejguan** [comment]
@PhilMacKay 
This is a good question. We use `Sequence` because it matches our minimum abstract method requirements (`__getitem__` and `__len__`). Those two methods are the subset for `Mapping` as well. You are right that this might not be that obvious for our users. I am fine to add a `CollectionWrapper` to `torchdata.datapipe.map`. WDYT? @NivekT 

Besides, in order to avoid BC breaking, I prefer aliasing option.

**Comment by NivekT** [comment]
I think `Sequence` is [correct](https://docs.python.org/3/library/collections.abc.html) but we can make an alias `MappingWrapper`? I am open to `CollectionWrapper` as well but is that too broad?

**Comment by JanFidor** [comment]
Could I pick up Form, Demux or  Multiplexer? Also my final exams are coming soon, so I'll most likely start working on it in like 2-3 weeks tops if it's not a problem

**Comment by NivekT** [comment]
Hi @JanFidor,

To avoid implementing every single feature twice (one in Iter-style, one in Map-style), our current [guidance](https://pytorch.org/data/main/torchdata.datapipes.map.html) is that users should use the built-in `IterDataPipe` for various functionalities, and convert it to MapDataPipe as needed using [`IterToMapConverter`](https://pytorch.org/data/main/generated/torchdata.datapipes.map.IterToMapConverter.html#torchdata.datapipes.map.IterToMapConverter) or .to_map_datapipe(). For that reason, I don't think we will need additional `MapDataPipe` for now.

`Multiplexer` is do-able, we can add it if we think it has sufficient value. @ejguan What do you think?

**Comment by josiahls** [comment]
@JanFidor I've was originally going to do a pr, but haven't had the time. I have an implementation of mux / demux for map datapipes:

https://github.com/josiahls/fastrl/blob/feature/trpo/fastrl/pipes/map/mux.py
https://github.com/josiahls/fastrl/blob/feature/trpo/fastrl/pipes/map/demux.py

Feel free to use those as reference, make your own, and make a pr. I would do it myself, but this issue has been open for a while, so I'm happy if someone else wants to get an actual pr made for a map mux/demux implimentation. 

These probably would need unit tests also, which I'm missing.

**Comment by devpramod-intel** [comment]
@josiahls I can make a PR for this.

**Comment by devpramod** [comment]
@josiahls I can make a PR for this.

**Comment by NivekT** [comment]
Hi, we will accept a PR if there is an actual use case for the functionality. Otherwise, we do not plan to add more DataPipes.

We are definitely open to suggestions so let us know what the use case will be!

**Comment by NivekT** [comment]
Yes, I can close it. I do think we have clear guidance about `MapDataPipe` and its usage. We likely won't add many more in the future but we will stay open to suggestions and consider new use cases as they come up.
